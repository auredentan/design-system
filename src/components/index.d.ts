export type Color = 'primary' | 'blue' | 'green' | 'yellow' | 'red' | string;
export type Emphasis = 'high' | 'medium' | 'low';
export type Theme = 'light' | 'dark';
export type Variant = 'button' | 'icon';
export type Size = 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl';
