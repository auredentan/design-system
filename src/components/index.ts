/////////////////////////////
//                         //
//    Public attributes    //
//                         //
/////////////////////////////

/**
 * Defines the type of a complex default prop (which depends on the value of another prop).
 */
interface IComplexPropDefault<T> {
    default: T;
    [key: string]: T;
}
type ComplexPropDefault<T> = IComplexPropDefault<T>;

enum Colors {
    primary = 'primary',
    blue = 'blue',
    dark = 'dark',
    green = 'green',
    yellow = 'yellow',
    red = 'red',
}
type Color = Colors | string;

enum Themes {
    light = 'light',
    dark = 'dark',
}
type Theme = Themes;

enum Sizes {
    xxs = 'xxs',
    xs = 'xs',
    s = 's',
    m = 'm',
    l = 'l',
    xl = 'xl',
    xxl = 'xxl',
}
type Size = Sizes;

/////////////////////////////

export { ComplexPropDefault, Color, Colors, Theme, Themes, Size, Sizes };
