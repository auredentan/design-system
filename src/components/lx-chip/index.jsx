export { LxChip } from './react/LxChip';
