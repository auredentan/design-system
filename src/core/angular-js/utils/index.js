import './lx-depth_service';
import './lx-enter-keypress';
import './lx-event-scheduler_service';
import './lx-utils_service';
