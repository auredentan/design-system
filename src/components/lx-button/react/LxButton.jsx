import React from 'react';
import classNames from 'classnames';

import { handleBasicClasses } from '../../../core/utils';
import '../style/lx-button.scss';

const LxButtonRoot = ({ href, children, ...props }) => {
    return href ? (
        <a href={href} {...props}>
            {children}
        </a>
    ) : (
        <button {...props}>{children}</button>
    );
};

const LxButton = ({
    children,
    className,
    color = 'primary',
    variant = 'primary',
    size = 'm',
    theme = 'light',
    ...props
}) => (
    <LxButtonRoot
        className={classNames(className, handleBasicClasses({ color, size, theme, variant, prefix: 'lx-button' }))}
        {...props}
    >
        {children}
    </LxButtonRoot>
);

export { LxButton };
