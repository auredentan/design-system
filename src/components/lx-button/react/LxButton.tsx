import { Color, ComplexPropDefault, Emphasis, Size, Theme, Variant } from 'components';

import { LxButtonRootProps } from './LxButtonRoot';

/////////////////////////////

import React, { Children } from 'react';

import classNames from 'classnames';

import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';
import isString from 'lodash/isString';

import { LxIcon } from 'LumX';
import { isElementOfType, isElementText, unwrapFragment } from 'LumX/core/react/utils';
import { handleBasicClasses } from 'LumX/core/utils';

import { LxButtonRoot } from './LxButtonRoot';

/////////////////////////////

/**
 * The default class name and classes prefix for this component.
 *
 * @type {string}
 * @constant
 * @readonly
 */
const CLASSNAME: string = 'lx-button';

/////////////////////////////

/**
 * Defines the props of the <LxButton> component.
 */
interface ILxButtonProps {
    /**
     * The button color which must be defined by `lx-button--${color}` css class.
     */
    color?: Color;

    /**
     * The emphasis of the button which must be defined by `lx-button-emphasis--${emphasis}` css class.
     */
    emphasis?: Emphasis;

    /**
     * The button size which must be defined by `lx-button--${size}` css class.
     */
    size?: Size;

    /**
     * The button theme which must be defined by `lx-button--${theme}` css class.
     */
    theme?: Theme;

    /**
     * The button variant which must be defined by `lx-type--${variant}` css class.
     */
    variant?: Variant;
}
type LxButtonProps = ILxButtonProps & LxButtonRootProps;

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Validate the <LxButton> component props and children.
 * Also, sanitize, cleanup and format the children and return the processed ones.
 *
 * @param  {LxButtonProps}   props The children and props of the <LxButton> component.
 * @return {React.ReactNode} The processed children of the component.
 */
function _validate({ children, variant }: LxButtonProps): React.ReactNode {
    let newChildren: React.ReactNode = children;

    let childrenCount: number = Children.count(newChildren);
    if (childrenCount === 0) {
        throw new Error('Your <LxButton> must have at least 1 child for the label or the icon (got 0)!');
    }

    newChildren = unwrapFragment(newChildren);
    childrenCount = Children.count(newChildren);

    if (variant === 'button') {
        if (childrenCount > 3) {
            throw new Error(
                `You cannot have more than 3 children (an icon, a label, another icon) in a 'button' \`variant\` of <LxButton> (got ${childrenCount})!`,
            );
        }

        let index: number = -1;
        const childrenTypes: string[] = [];
        newChildren = Children.map(
            newChildren,
            (child: any): React.ReactElement => {
                index++;

                let newChild: React.ReactElement = child;
                if (isString(child)) {
                    newChild = <span>{child}</span>;
                }

                const isChildText: boolean = isElementText(newChild);
                const isChildIcon: boolean = isElementOfType(newChild, LxIcon);

                if (!isChildText && !isChildIcon) {
                    throw new Error(
                        `The children of a <LxButton> must be either of type text, <span> or <LxIcon> (got '${get(
                            newChild.type,
                            'name',
                            newChild.type,
                        ) || `'${child}'`}}')!`,
                    );
                }

                childrenTypes[index] = isChildText ? 'text' : 'icon';

                if (isChildText && index > 0 && childrenTypes[index - 1] === 'text') {
                    throw new Error(
                        `You cannot have more than 1 text or <span> child (for the label) in a 'button' \`variant\` of <LxButton> (got at least 2)!`,
                    );
                }

                if (isChildIcon) {
                    if (childrenCount === 1) {
                        console.warn(
                            "If you want to display an icon button, you should use the 'icon' `variant` of the <LxButton> instead of the 'button' `variant`\nYou should even consider using the <LxIconButton> component instead.",
                        );
                    }

                    if (index > 0 && childrenTypes[index - 1] === 'icon') {
                        throw new Error(
                            `You cannot have 2 icons children following in a 'button' \`variant\` of <LxButton>, there must be 1 text or <span> children (for the label) between them!`,
                        );
                    }
                }

                return newChild;
            },
        );
    } else if (variant === 'icon') {
        if (childrenCount > 1) {
            throw new Error(
                `You cannot have more than 1 child in an 'icon' \`variant\` of <LxButton> (got ${childrenCount})!`,
            );
        }

        Children.forEach(
            newChildren,
            (child: any): void => {
                if (isElementOfType(child, LxIcon)) {
                    return;
                }

                throw new Error(
                    `You can only have a <LxIcon> child in an 'icon' \`variant\` of <LxButton> (got ${get(
                        child.type,
                        'name',
                        child.type,
                    ) || `'${child}'`})!`,
                );
            },
        );
    }

    return newChildren;
}

/////////////////////////////

/**
 * Define the types of the default props.
 */
interface ILxButtonDefaultPropsType {
    color: ComplexPropDefault<Color>;
    emphasis: Emphasis;
    size: Size;
    theme: Theme;
    variant: Variant;
}

/**
 * The default value of props.
 *
 * @type {ILxButtonDefaultPropsType}
 * @constant
 * @readonly
 */
const DEFAULT_PROPS: ILxButtonDefaultPropsType = {
    color: {
        default: 'dark' as Color,
        'emphasis-high': 'primary' as Color,
    },
    emphasis: 'high' as Emphasis,
    size: 'm' as Size,
    theme: 'light' as Theme,
    variant: 'button' as Variant,
};

/**
 * Displays a button.
 * If the `href` property is set, it will display a `<a>` HTML tag. If not, it will use a `<button>` HTML tag instead.
 *
 * @return {JSX.Element} The <LxButton> component.
 */
const LxButton: React.FC<LxButtonProps> = ({
    children,
    className = '',
    color,
    emphasis = DEFAULT_PROPS.emphasis,
    size = DEFAULT_PROPS.size,
    theme = DEFAULT_PROPS.theme,
    variant = DEFAULT_PROPS.variant,
    ...props
}: LxButtonProps): JSX.Element => {
    children = _validate({ children, className, color, emphasis, size, theme, variant, ...props });

    if (variant === 'button') {
        let index = -1;
        Children.forEach(
            children,
            (child: any): void => {
                index++;

                if (isElementOfType(child, LxIcon)) {
                    className += isEmpty(className) ? '' : ' ';
                    className += `${CLASSNAME}--has-${index === 0 ? 'left' : 'right'}-icon`;
                }
            },
        );
    }

    if (isEmpty(color)) {
        color = DEFAULT_PROPS.color[`emphasis-${emphasis}`] || DEFAULT_PROPS.color.default;
    }

    return (
        <LxButtonRoot
            className={classNames(
                className,
                handleBasicClasses({ prefix: CLASSNAME, color, emphasis, size, theme, variant }),
            )}
            {...props}
        >
            {children}
        </LxButtonRoot>
    );
};
LxButton.displayName = 'LxButton';

/////////////////////////////

export { CLASSNAME, DEFAULT_PROPS, LxButton, LxButtonProps };
