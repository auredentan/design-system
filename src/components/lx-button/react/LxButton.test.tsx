import { ICommonSetup } from 'LumX/core/testing/utils.test';

import { Color, Emphasis, Size, Theme, Variant } from 'components';

import { LxButtonProps } from './LxButton';

/////////////////////////////

import React, { Fragment } from 'react';

import { shallow, ShallowWrapper } from 'enzyme';

import get from 'lodash/get';

import { LxButton, LxIcon } from 'LumX';
import { getBasicClass } from 'LumX/core/utils';
import { mdiPlus } from 'LumX/icons';

import { CLASSNAME, DEFAULT_PROPS } from './LxButton';

/////////////////////////////

/**
 * Define the overriding properties waited by the `setup` function.
 */
interface ISetupProps extends Partial<LxButtonProps> {}

/**
 * Defines what the `setup` function will return.
 */
interface ISetup extends ICommonSetup {
    props: ISetupProps;

    /**
     * The <LxButtonRoot> element that is used as a wrapper for the children of the <LxButton>.
     */
    buttonRoot: ShallowWrapper;

    /**
     * The <LxIcon> icon(s) of the <LxButton> (can have 0, 1 or 2).
     */
    icon: ShallowWrapper;

    /**
     * The <span> element of label of the <LxButton> (can have 0 or 1).
     */
    label: ShallowWrapper;
}

/////////////////////////////
//                         //
//    Private functions    //
//                         //
/////////////////////////////

/**
 * Get the default value of the given prop of a <LxButton>, depending on the effective props of the component (some
 * default value depends on the value of another prop).
 *
 * @param {string}      prop  The name of the prop you want the default value of.
 * @param {ISetupProps} props The current props of the <LxButton>.
 */
function _getDefaultPropValue({ prop, props }: { prop: string; props?: ISetupProps }): string {
    return prop === 'color'
        ? DEFAULT_PROPS[prop][`emphasis-${get(props, 'emphasis', DEFAULT_PROPS.emphasis)}`] ||
              DEFAULT_PROPS[prop].default
        : DEFAULT_PROPS[prop];
}

/////////////////////////////

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 *
 * @param  {ISetupProps} props  The props to use to override the default props of the component.
 * @return {ISetup}      An object with the props, the component wrapper and some shortcut to some element inside of the
 *                       component.
 */
const setup = ({ ...propsOverrides }: ISetupProps = {}): ISetup => {
    const props: LxButtonProps = {
        children: 'Label',
        ...propsOverrides,
    };

    const wrapper: ShallowWrapper = shallow(<LxButton {...props} />);

    return {
        buttonRoot: wrapper.find('LxButtonRoot'),

        icon: wrapper.find('LxIcon'),
        label: wrapper.find('span'),

        props,
        wrapper,
    };
};

describe(`<${LxButton.displayName}>`, (): void => {
    // 1. Test render via snapshot (default states of component).
    describe('Snapshots and structure', (): void => {
        it('should render correctly a text label as a button', (): void => {
            const { icon, label, wrapper }: ISetup = setup();
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(false);
            expect(label.exists()).toEqual(true);
        });

        it('should render correctly a <span> label as a button', (): void => {
            const children: React.ReactNode = <span>Label</span>;

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(false);
            expect(label.exists()).toEqual(true);
        });

        it('should render correctly an icon and a text label as a button', (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    Label
                </Fragment>
            );

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(1);
            expect(label.exists()).toEqual(true);
        });

        it('should render correctly an icon and a <span> label as a button', (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <span>Label</span>
                </Fragment>
            );

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(1);
            expect(label.exists()).toEqual(true);
        });

        it('should render correctly a text label and an icon as a button', (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    Label
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(1);
            expect(label.exists()).toEqual(true);
        });

        it('should render correctly a <span> label and an icon as a button', (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <span>Label</span>
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(1);
            expect(label.exists()).toEqual(true);
        });

        it('should render correctly two icons and a text label as a button', (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    Label
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(2);
            expect(label.exists()).toEqual(true);
        });

        it('should render correctly two icons and a <span> label as a button', (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <span>Label</span>
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(2);
            expect(label.exists()).toEqual(true);
        });

        it("should render correctly an icon button with the 'button' `variant` as a button", (): void => {
            // Disable the display of the warn message in the console.
            global.console.warn = jest.fn();

            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            const { icon, label, wrapper }: ISetup = setup({ children });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(1);
            expect(label.exists()).toEqual(false);

            // @ts-ignore
            global.console.warn.mockClear();
        });

        it("should render correctly an icon button with the 'icon' `variant` as a button", (): void => {
            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            const { icon, label, wrapper }: ISetup = setup({ children, variant: 'icon' });
            expect(wrapper).toMatchSnapshot();

            expect(icon.exists()).toEqual(true);
            expect(icon.length).toEqual(1);
            expect(label.exists()).toEqual(false);
        });
    });

    /////////////////////////////

    // 2. Test defaultProps value and important props custom values.

    describe('Props', (): void => {
        it('should use default props', (): void => {
            const { buttonRoot }: ISetup = setup();

            Object.keys(DEFAULT_PROPS).forEach(
                (prop: string): void => {
                    const defaultValue: string = _getDefaultPropValue({ prop });

                    expect(
                        buttonRoot.hasClass(getBasicClass({ prefix: CLASSNAME, type: prop, value: defaultValue })),
                    ).toEqual(true);
                },
            );
        });

        it('should use the given `color`', (): void => {
            const testedProp: string = 'color';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'green' as Color,
            };

            const { buttonRoot }: ISetup = setup(modifiedProps);

            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
                ),
            ).toEqual(true);
            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: DEFAULT_PROPS[testedProp] }),
                ),
            ).toEqual(false);
        });

        it('should use the given `emphasis`', (): void => {
            const testedProp: string = 'emphasis';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'low' as Emphasis,
            };

            const { buttonRoot }: ISetup = setup(modifiedProps);

            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
                ),
            ).toEqual(true);
            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: DEFAULT_PROPS[testedProp] }),
                ),
            ).toEqual(false);
        });

        it('should use the given `size`', (): void => {
            const testedProp: string = 'size';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'xxs' as Size,
            };

            const { buttonRoot }: ISetup = setup(modifiedProps);

            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
                ),
            ).toEqual(true);
            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: DEFAULT_PROPS[testedProp] }),
                ),
            ).toEqual(false);
        });

        it('should use the given `theme`', (): void => {
            const testedProp: string = 'theme';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'dark' as Theme,
            };

            const { buttonRoot }: ISetup = setup(modifiedProps);

            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
                ),
            ).toEqual(true);
            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: DEFAULT_PROPS[testedProp] }),
                ),
            ).toEqual(false);
        });

        it('should use the given `variant`', (): void => {
            const testedProp: string = 'variant';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'icon' as Variant,
                children: <LxIcon icon={mdiPlus} />,
            };

            const { buttonRoot }: ISetup = setup(modifiedProps);

            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: modifiedProps[testedProp] }),
                ),
            ).toEqual(true);
            expect(
                buttonRoot.hasClass(
                    getBasicClass({ prefix: CLASSNAME, type: testedProp, value: DEFAULT_PROPS[testedProp] }),
                ),
            ).toEqual(false);
        });

        it('should forward any CSS class', (): void => {
            const testedProp: string = 'className';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'component component--is-tested',
            };

            const { buttonRoot }: ISetup = setup(modifiedProps);

            expect(buttonRoot.prop(testedProp)).toContain(modifiedProps[testedProp]);
        });

        it('should forward any other props', (): void => {
            const testedProp: string = 'winter';
            const modifiedProps: ISetupProps = {
                [testedProp]: 'is coming',
            };

            const { buttonRoot }: ISetup = setup(modifiedProps);

            expect(buttonRoot.prop(testedProp)).toEqual(modifiedProps[testedProp]);
        });
    });

    /////////////////////////////

    // 3. Test events.

    /////////////////////////////

    // 4. Test conditions (i.e. things that display or not in the UI based on props).
    describe('Conditions', (): void => {
        beforeEach(
            (): void => {
                try {
                    // If `console.warn` has been mocked at least one, this exists. So disable TS here.
                    // @ts-ignore
                    global.console.warn.mockClear();
                } catch (exception) {
                    // Nothing to do here.
                }
            },
        );

        it('should fail when no child is given', (): void => {
            expect(
                (): void => {
                    setup({ children: null });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should fail when more than 3 children are given in the 'button' `variant`", (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <span>label</span>
                    <span>label 2</span>
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            expect(
                (): void => {
                    // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                    // @ts-ignore
                    setup({ children });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when anything else than a text, <span> or <${
            LxIcon.displayName
        }> is given as child in the 'button' \`variant\``, (): void => {
            const children: React.ReactNode = <div>toto</div>;

            expect(
                (): void => {
                    setup({ children });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should fail when more than 1 text or <span> child is given in the 'button' `variant`", (): void => {
            let children: React.ReactNode = (
                <Fragment>
                    <span>label</span>
                    label 2
                </Fragment>
            );

            expect(
                (): void => {
                    // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                    // @ts-ignore
                    setup({ children });
                },
            ).toThrowErrorMatchingSnapshot();

            /////////////////////////////

            children = (
                <Fragment>
                    <span>label</span>
                    <span>label 2</span>
                </Fragment>
            );

            expect(
                (): void => {
                    // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                    // @ts-ignore
                    setup({ children });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when there are 2 or more <${
            LxIcon.displayName
        }> in a row in the 'button' \`variant\``, (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            expect(
                (): void => {
                    // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                    // @ts-ignore
                    setup({ children });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should fail when more than 1 child is given in the 'icon' `variant`", (): void => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            expect(
                (): void => {
                    // We know that children is mandatory, but we want to test the error validation. So disable TS here.
                    // @ts-ignore
                    setup({ children, variant: 'icon' });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it(`should fail when anything else than a <${
            LxIcon.displayName
        }> is given as child in the 'icon' \`variant\``, (): void => {
            const children: React.ReactNode = 'Toto';

            expect(
                (): void => {
                    setup({ children, variant: 'icon' });
                },
            ).toThrowErrorMatchingSnapshot();
        });

        it("should warn the user when rendering an icon button with the 'button' `variant`", (): void => {
            global.console.warn = jest.fn();

            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            setup({ children });
            expect(global.console.warn).toHaveBeenCalled();
        });

        it("should not warn the user when rendering an icon button with the 'icon' `variant`", (): void => {
            global.console.warn = jest.fn();

            const children: React.ReactNode = <LxIcon icon={mdiPlus} />;

            setup({ children, variant: 'icon' });
            expect(global.console.warn).not.toHaveBeenCalled();
        });

        it("should have no `theme` in any other `emphasis` than 'high'", (): void => {
            let modifiedProps: ISetupProps = {
                emphasis: 'high' as Emphasis,
            };

            let { buttonRoot }: ISetup = setup(modifiedProps);
            expect(
                buttonRoot.hasClass(getBasicClass({ prefix: CLASSNAME, type: 'theme', value: DEFAULT_PROPS.theme })),
            ).toEqual(true);

            /////////////////////////////

            modifiedProps = {
                emphasis: 'medium' as Emphasis,
            };

            ({ buttonRoot } = setup(modifiedProps));
            expect(
                buttonRoot.hasClass(getBasicClass({ prefix: CLASSNAME, type: 'theme', value: DEFAULT_PROPS.theme })),
            ).toEqual(false);

            /////////////////////////////

            modifiedProps = {
                emphasis: 'low' as Emphasis,
            };

            ({ buttonRoot } = setup(modifiedProps));
            expect(
                buttonRoot.hasClass(getBasicClass({ prefix: CLASSNAME, type: 'theme', value: DEFAULT_PROPS.theme })),
            ).toEqual(false);
        });

        it('should only have the "left-icon" CSS class when an icon is passed as first child of a \'button\' `variant`', () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    Label
                </Fragment>
            );

            const { buttonRoot }: ISetup = setup({ children });
            expect(buttonRoot.hasClass(`${CLASSNAME}--has-left-icon`)).toEqual(true);
            expect(buttonRoot.hasClass(`${CLASSNAME}--has-right-icon`)).toEqual(false);
        });

        it('should only have the "right-icon" CSS class when an icon is passed as last child of a \'button\' `variant`', () => {
            const children: React.ReactNode = (
                <Fragment>
                    Label
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { buttonRoot }: ISetup = setup({ children });
            expect(buttonRoot.hasClass(`${CLASSNAME}--has-left-icon`)).toEqual(false);
            expect(buttonRoot.hasClass(`${CLASSNAME}--has-right-icon`)).toEqual(true);
        });

        it('should have both "left-icon" and "right-icon" CSS classes when icons are passed as first and last children of a \'button\' `variant`', () => {
            const children: React.ReactNode = (
                <Fragment>
                    <LxIcon icon={mdiPlus} />
                    Label
                    <LxIcon icon={mdiPlus} />
                </Fragment>
            );

            const { buttonRoot }: ISetup = setup({ children });
            expect(buttonRoot.hasClass(`${CLASSNAME}--has-left-icon`)).toEqual(true);
            expect(buttonRoot.hasClass(`${CLASSNAME}--has-right-icon`)).toEqual(true);
        });
    });

    /////////////////////////////

    // 5. Test state.
});
