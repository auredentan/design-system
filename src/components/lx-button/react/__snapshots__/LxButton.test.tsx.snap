// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LxButton> Conditions should fail when anything else than a <LxIcon> is given as child in the 'icon' \`variant\` 1`] = `"You can only have a <LxIcon> child in an 'icon' \`variant\` of <LxButton> (got 'Toto')!"`;

exports[`<LxButton> Conditions should fail when anything else than a text, <span> or <LxIcon> is given as child in the 'button' \`variant\` 1`] = `"The children of a <LxButton> must be either of type text, <span> or <LxIcon> (got 'div}')!"`;

exports[`<LxButton> Conditions should fail when more than 1 child is given in the 'icon' \`variant\` 1`] = `"You cannot have more than 1 child in an 'icon' \`variant\` of <LxButton> (got 2)!"`;

exports[`<LxButton> Conditions should fail when more than 1 text or <span> child is given in the 'button' \`variant\` 1`] = `"You cannot have more than 1 text or <span> child (for the label) in a 'button' \`variant\` of <LxButton> (got at least 2)!"`;

exports[`<LxButton> Conditions should fail when more than 1 text or <span> child is given in the 'button' \`variant\` 2`] = `"You cannot have more than 1 text or <span> child (for the label) in a 'button' \`variant\` of <LxButton> (got at least 2)!"`;

exports[`<LxButton> Conditions should fail when more than 3 children are given in the 'button' \`variant\` 1`] = `"You cannot have more than 3 children (an icon, a label, another icon) in a 'button' \`variant\` of <LxButton> (got 4)!"`;

exports[`<LxButton> Conditions should fail when no child is given 1`] = `"Your <LxButton> must have at least 1 child for the label or the icon (got 0)!"`;

exports[`<LxButton> Conditions should fail when there are 2 or more <LxIcon> in a row in the 'button' \`variant\` 1`] = `"You cannot have 2 icons children following in a 'button' \`variant\` of <LxButton>, there must be 1 text or <span> children (for the label) between them!"`;

exports[`<LxButton> Snapshots and structure should render correctly a <span> label 1`] = `
<LxButtonRoot
  className="lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly a <span> label and an icon 1`] = `
<LxButtonRoot
  className="lx-button--has-right-icon lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".1"
  />
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly a text label 1`] = `
<LxButtonRoot
  className="lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly a text label and an icon 1`] = `
<LxButtonRoot
  className="lx-button--has-right-icon lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <span
    key=".0"
  >
    Label
  </span>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".1"
  />
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly an icon and a <span> label 1`] = `
<LxButtonRoot
  className="lx-button--has-left-icon lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly an icon and a text label 1`] = `
<LxButtonRoot
  className="lx-button--has-left-icon lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly an icon button with the 'button' \`variant\` 1`] = `
<LxButtonRoot
  className="lx-button--has-left-icon lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly an icon button with the 'icon' \`variant\` 1`] = `
<LxButtonRoot
  className="lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-icon"
>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
  />
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly two icons and a <span> label 1`] = `
<LxButtonRoot
  className="lx-button--has-left-icon lx-button--has-right-icon lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".2"
  />
</LxButtonRoot>
`;

exports[`<LxButton> Snapshots and structure should render correctly two icons and a text label 1`] = `
<LxButtonRoot
  className="lx-button--has-left-icon lx-button--has-right-icon lx-button lx-button--color-primary lx-button--emphasis-high lx-button--size-m lx-button--theme-light lx-button--variant-button"
>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".0"
  />
  <span
    key=".1"
  >
    Label
  </span>
  <LxIcon
    icon="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
    key=".2"
  />
</LxButtonRoot>
`;
